infrastructure/src/main/java/org/corfudb/infrastructure/RequestHandlerMethods.java

```diff
@@ -130,10 +133,50 @@ private void registerMethod(@Nonnull final MethodHandles.Lookup caller,
                         mtt, mh, mtt).getTarget().bindTo(server).invoke();
             }
 
-            final HandlerMethod handler = h::handle;
+            // Install pre-conditions on handler
+            final HandlerMethod handler = generateConditionalHandler(annotation.type(), h);
+            // Install the handler in the map
             handlerMap.put(annotation.type(), handler);
         } catch (Throwable e) {
             throw new UnrecoverableCorfuError("Exception during request handler registration", e);
         }
     }
+
+    /** Generate a conditional handler, which instruments the handler with metrics if configured
+     * as enabled and checks whether the server is shutdown and ready.
+     * @param type          The type the request message is being handled for.
+     * @param handler       The {@link RequestHandlerMethods.HandlerMethod} which handles the message.
+     * @return              A new {@link RequestHandlerMethods.HandlerMethod} which conditionally executes the
+     *                      handler based on preconditions (whether the server is shutdown/ready).
+     */
+    private HandlerMethod generateConditionalHandler(@NonNull final PayloadCase type,
+                                                     @NonNull final HandlerMethod handler) {
```

Comment:
nit: the javadoc's indent is broken here

Reply:
Fixed.